### ISSUE TITLE
missing dependencies on wordpress 6.4

### ISSUE THREAD

**Issue by signsi** [studiopress/genesis-custom-blocks#168]
## Bug summary

On wordpress 6.4 our custom blocks based on genesis are no longer displayed in the editor. 

1. The first issues is that [wordpress dropped lodash](https://github.com/WordPress/gutenberg/pull/52571) but genesis seems to need it, possibly without properly requesting it.
2. The second issue is: `TypeError: undefined is not an object (evaluating 'rn.debounce')`.

The first issues can be quickfixed by explicitly loading lodash for example a script which is loaded in the editor. We are currently still looking for quickfixes for the second issue.

## Steps to reproduce

1. Update wordpress to 6.4
2. Check genesis custom blocks in the editor
3. check the browser console for errors

## Expected behavior
The Plugin should properly work with all stable version of wordpress.
## Actual behavior
Multiple js errors. On the frontend side it still seems to work.

## JS Console

```
TypeError: undefined is not an object (evaluating 'rn.debounce')
value — block-editor.js:2:71662
Jr — react-dom.min.js:4273
Zr — react-dom.min.js:4254:94
Gr — react-dom.min.js:4227
(anonymous function) — react-dom.min.js:5027
xl — react-dom.min.js:5051
fl — react-dom.min.js:4630:82
Nn — react-dom.min.js:1856
(anonymous function) — react-dom.min.js:4462
```
## Suspected cause (optional)
It looks like a missing functions (debounce). Possibly it is due to a version conflict of react dom.

## Versions

 - WordPress version: 6.4
 - Genesis Custom Blocks version:  1.5.3 via wpmudev / 1.5.2 via github (both don't work, why is this discrepancy?)
 - OS: Mac
 - Browser: Safari
 - Device: MBP

**Comment by kienstra** [comment]
Hi @signsi,
Thanks for letting me know about this.

It sounds like there aren't any other plugins active on your site, right?

I'm not seeing this error, and the [e2e tests are passing with 6.4](https://app.circleci.com/pipelines/github/studiopress/genesis-custom-blocks/1746/workflows/02124566-0585-4453-ace8-a8df56888cff/jobs/14637?invite=true#step-117-1475_47).

**Comment by kienstra** [comment]
PS. Thanks for your very detailed report.

**Comment by kienstra** [comment]
> Genesis Custom Blocks version: 1.5.3 via wpmudev / 1.5.2 via github (both don't work, why is this discrepancy?)

Good question, we've [never](https://plugins.trac.wordpress.org/browser/genesis-custom-blocks/tags) [released](https://github.com/studiopress/genesis-custom-blocks/tags) a version `1.5.3`, that's strange.

**Comment by signsi** [comment]
Hey @kienstra 

I've also informed wpengine. We have version 1.5.3 from there. What is the connection there? The error exists with all versions. I also have the error with the default theme and without other plugins.

Regarding lodash, the clean solution is probably that the dependency is explictly specified. Lodash is still available [in the core](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L113), but is no longer automatically enqueued. 

I've just downloaded the version from the repo here. So far updates of the wordpress / react dependencies have not helped so far.

**Comment by signsi** [comment]
Quick update: The problem with lodash seems to be due to a combination of genesis custom blocks and our theme (since 6.4). However, the problem with the debounce always exists.

**Comment by kienstra** [comment]
>I've also informed wpengine. We have version 1.5.3 from there. What is the connection there?

Ah, so that's Genesis Custom Blocks Pro. This repo is for the free version [deployed to wp.org](https://wordpress.org/plugins/genesis-custom-blocks/).

Genesis Custom Blocks Pro also runs GCB free, though.

**Comment by kienstra** [comment]
>Regarding lodash, the clean solution is probably that the dependency is explictly specified. Lodash is still available [in the core](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L113), but is no longer automatically enqueued.

Hm, GCB [enqueues lodash](https://plugins.trac.wordpress.org/browser/genesis-custom-blocks/trunk/js/dist/block-editor.asset.php?rev=2946749), I'm not sure what more we can do with that.

**Comment by signsi** [comment]
> > Regarding lodash, the clean solution is probably that the dependency is explictly specified. Lodash is still available [in the core](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L113), but is no longer automatically enqueued.
> 
> Hm, GCB [enqueues lodash](https://plugins.trac.wordpress.org/browser/genesis-custom-blocks/trunk/js/dist/block-editor.asset.php?rev=2946749), I'm not sure what more we can do with that.

That issue was not related to the plugin.

**Comment by kienstra** [comment]
Ah, thanks.

**Comment by kienstra** [comment]
I'm [looking at rn.debounce](https://plugins.trac.wordpress.org/browser/genesis-custom-blocks/trunk/js/dist/block-editor.js?rev=2946749), it's hard to see where that's coming from.

**Comment by signsi** [comment]
I was also able to reproduce the error on a bare WP installation on a standard web server. So I'm pretty sure it's due to the combination of WP 6.4 and Genesis Custom Blocks.
```
TypeError: undefined is not an object (evaluating 'rn.debounce')
value — block-editor.js:2:71662
Jr — react-dom.min.js:10:80732
Zr — react-dom.min.js:10:80492
Gr — react-dom.min.js:10:80026
(anonymous function) — react-dom.min.js:10:91529
xl — react-dom.min.js:10:92040
fl — react-dom.min.js:10:85811
Nn — react-dom.min.js:10:32476
(anonymous function) — react-dom.min.js:10:83229
```

**Comment by kienstra** [comment]
OK, thanks. I don't see that error. 

Would you be able to run this plugin from source to see if that gives a better error message?

```
git clone https://github.com/studiopress/genesis-custom-blocks
cd genesis-custom-blocks
nvm use && npm ci && npm run dev
```

**Comment by signsi** [comment]
> I'm [looking at rn.debounce](https://plugins.trac.wordpress.org/browser/genesis-custom-blocks/trunk/js/dist/block-editor.js?rev=2946749), it's hard to see where that's coming from.

I totally agree. Tried it some hours ago myself.
Not sure if this helps but: When I create a new page and insert the block, everything seems to be correct. Only when I click out or save does the error appear.

**Comment by kienstra** [comment]
>Only when I click out or save does the error appear.

OK, thanks. I'll do that.

**Comment by kienstra** [comment]
Sorry, still not seeing the error. I'll think about this today.

**Comment by signsi** [comment]
I tried the same thing today. But I used the build instead of the dev server.
So the plugin runs from the source code, in dev mode.  The error has changed. We are back to lodash.
```
TypeError: undefined is not an object (evaluating 'lodash__WEBPACK_IMPORTED_MODULE_8__.debounce')
componentDidMount — server-side-render.js:64
Jr — react-dom.min.js:10:80732
Zr — react-dom.min.js:10:80492
Gr — react-dom.min.js:10:80026
(anonymous function) — react-dom.min.js:10:91529
xl — react-dom.min.js:10:92040
fl — react-dom.min.js:10:85811
Nn — react-dom.min.js:10:32476
(anonymous function) — react-dom.min.js:10:83229
```

**Comment by signsi** [comment]
[Here](https://github.com/WordPress/gutenberg/blob/%40wordpress/server-side-render%401.19.5/packages/server-side-render/src/server-side-render.js#L36) you will find the source code. It is from the package @wordpress/server-side-render in version 1.19.5.

**Comment by kienstra** [comment]
Wow, great find.

So [debounce](https://github.com/WordPress/gutenberg/blob/%40wordpress/server-side-render%401.19.5/packages/server-side-render/src/server-side-render.js#L4) is `undefined`.

**Comment by signsi** [comment]
Yes, that's how it looks at the moment. I'll take another deep dive into the matter tomorrow.
Thanks Ryan!

**Comment by kienstra** [comment]
Thanks, talk to you tomorrow!

**Comment by kienstra** [comment]
The e2e tests are [showing the same](https://app.circleci.com/pipelines/github/studiopress/genesis-custom-blocks/1747/workflows/ff7d3f10-f39e-4734-83fb-c73928fbfa47/jobs/14648) error about `debounce`.

**Comment by signsi** [comment]
Hey Ryan
Great, so now we have confirmation that it's not just down to our setup. I'll get back to you later.
Thanks!

**Comment by signsi** [comment]
We've just discussed the issue internally. This warning could potentially be related as the [refactoring to hooks](https://github.com/WordPress/gutenberg/commit/09671e394c452046a58894c60fc24f36bd24a2fe) also concerned these (or at least related) packages: 
`wp.compose.withState is deprecated since version 5.8. Please use wp.element.useState instead.`

**Comment by signsi** [comment]
To be more precise: the warning clearly comes from wp core.

**Comment by signsi** [comment]
Work in progress: Testing [@wordpress/server-side-render@4.3.8](https://github.com/WordPress/gutenberg/blob/%40wordpress/server-side-render%404.3.8/packages/server-side-render/src/server-side-render.js). Currently it looks much better. Doing some proper testing now. In Version 4.3.8 the refactoring to hooks is present.

**Comment by fabiankaimer** [comment]
Hi guys,
just wanted to let you know that I experience the exact same issue as @signsi with Wordpress 6.4.1 and Genesis Custom Blocks 1.5.2:

![image](https://github.com/studiopress/genesis-custom-blocks/assets/15916113/04a48c7a-681b-4aae-9084-7bbaf70a55c8)

Downgrading Wordpress back to 3.6.x does resolve the issue temporarily. Thanks already to you for taking care of this!

**Comment by signsi** [comment]
Hello @fabiankaimer 
Thanks for letting us know. Our manual tests in Wordpress look good so far. I am working on running the e2e tests.

**Comment by kienstra** [comment]
Hi @signsi,
What does it show when you enter `lodash` in the JS console?

**Comment by kienstra** [comment]
Does this .zip fix it on your site?

[genesis-custom-blocks.1.5.2.zip](https://github.com/studiopress/genesis-custom-blocks/files/13309581/genesis-custom-blocks.1.5.2.zip)

It's from #170, which at least passes the e2e tests.

**Comment by vishu87** [comment]
Hi @kienstra 

I was also facing this issue, but replacing the plugin which this zip file removes the problem for me on WP 6.4.1

**Comment by kienstra** [comment]
@vishu87,
Thanks a lot for testing that, good to hear.

**Comment by signsi** [comment]
@kienstra same for me. ✅

**Comment by kienstra** [comment]
@signsi,
Great to hear, thanks a lot for testing it and reporting this.

**Comment by signsi** [comment]
Thanks, for the PR @kienstra . Thanks, for the PR.  So it was really missing JS dependencies, more precisely lodash?

**Comment by kienstra** [comment]
Yeah, it was [missing lodash](https://github.com/studiopress/genesis-custom-blocks/pull/170#pullrequestreview-1722745448), you called it 😄

**Comment by kienstra** [comment]
I might not release this fix to wp.org until next week, Thursday is getting a late for a release 😄

**Comment by signsi** [comment]
Sure, no worries. Thanks Ryan. For the pro-version it might take even more time, right?

**Comment by kienstra** [comment]
>For the pro-version it might take even more time, right?

No, I'll release the Pro version the same day as the free version.

**Comment by bringsonline** [comment]
Replacing the plugin with the one in the zip-file worked for me as well, problem solved. Thanks a lot @kienstra

**Comment by kienstra** [comment]
Thanks, @bringsonline! Great to hear.

**Comment by kienstra** [comment]
Hi @signsi,
The fix for this is released in Genesis Custom Blocks free and Pro `1.6.0`.

In `wp-admin/plugins/php`, you should see a notice soon to update your GCB Pro to `1.6.0`.

**Comment by fabiankaimer** [comment]
Thanks for the quick fix. Works again for me as well!

**Comment by kienstra** [comment]
Thanks, @fabiankaimer! Good to hear!

**Comment by signsi** [comment]
> Hi @signsi, The fix for this is released in Genesis Custom Blocks free and Pro `1.6.0`.
> 
> In `wp-admin/plugins/php`, you should see a notice soon to update your GCB Pro to `1.6.0`.

Great, thanks a lot.
